### ISSUE TITLE
Support for joins on multiple columns 

### ISSUE THREAD

**Issue by vishalhkd007** [vaexio/vaex#863]
I recently started looking into Vaex for performance reasons. Was looking to improve the data processing times compared to Pandas which I am currently using. Did not want to go the Spark route as that option is not available for the current project. The data I process using Pandas is sizable - over 5M records and over 200 columns.

Pandas is able to process most of the data, but it is slow. Loading the data takes around 30s. When switching to Vaex, the data loading time was an immediate improvement. Other statistical operations also showed considerable speed-ups.
The roadblock with Vaex was with support for multi-index and joins on multiple columns and pivot tables. This is a critical feature needed for my work since we are doing a ton of data slicing and dicing with groupings on multiple columns for data analytics and visualizations. Is there a way Vaex supports this?

**Comment by maartenbreddels** [comment]
Hi,

> multi-index

Why do you need this? Is this really important or do you need to get use to a different workflow? With large datasets, I hardly ever see a use-case where I need to lookup individual rows. If you miss it for other reasons, please explain.



> joins on multiple columns

This will be possible if we get deeper arrow integration, it depends a bit on funding for this as well.



> pivot tables

I am not sure we should try to support this. I am assuming you are using this with small amounts of data, after aggregation. A very valid workflow is bigdata -> vaex -> aggregate -> pandas.

cheers,

Maarten

**Comment by vishalhkd007** [comment]
Hi,



We deal with a lot of data for housing and finance industry where there is a constant need for aggregations using multiple data elements to identify trends and patterns across these groupings. For example, identifying house price indexing trends by neighborhoods and population groupings. Or number of house rentals by county and income groups etc.



Multi-index is the foundation in Pandas that allows for joins on multiple columns - where each join column becomes a layered index in the order of the join specified. This is quite valuable for the kind of analysis we run on our datasets.



As for pivot tables - they are not a necessity, but more like a nice-to-have feature. Ability to have support for joins on multiple elements with aggregations is more critical and a must have for our analysis.



Hope this helps.



Vishal Deshpande

Founder, MTekLabs

vishal@mteklabs.com​

http://www.mteklabs.com​

+1 - 540 454 8175




---- On Fri, 26 Jun 2020 08:44:24 -0500 Maarten Breddels <notifications@github.com> wrote ----




Hi,

multi-index


Why do you need this? Is this really important or do you need to get use to a different workflow? With large datasets, I hardly ever see a use-case where I need to lookup individual rows. If you miss it for other reasons, please explain.

joins on multiple columns


This will be possible if we get deeper arrow integration, it depends a bit on funding for this as well.

pivot tables


I am not sure we should try to support this. I am assuming you are using this with small amounts of data, after aggregation. A very valid workflow is bigdata -> vaex -> aggregate -> pandas.

cheers,

Maarten

**Comment by maartenbreddels** [comment]
Hi Vishal,

Ok, great explaination

> Multi-index is the foundation in Pandas that allows for joins on multiple columns

Multi-index and joining/groupby on multiple columns are two different things. I'd like to have support for joining on multiple columns, based on Apache Arrow. I cannot say when that will happen without explicit funding, but it's something we'd like to have for sure. I'm not convinced having a multi index (or a pandas-like index at all) is a benefit and see it as a separate feature (although they overlap a bit).

TLDR, Would really like to have join on multiple columns, but there is no direct funding for that atm.

Regards,

Maarten

**Comment by maartenbreddels** [comment]
FYI: groupby on multiple columns is supported.
